doc/developer/design/20320328_persist_columnar.md

```diff
@@ -0,0 +1,940 @@
+
+The Arrow spec does not include the concept of a Map but the `arrow2` and
+`arrow-rs` crates have a `MapArray` type that is a list of tuples.
+
+> **Alternative:** We could encode maps to some binary format, e.g. proto, and
+store them as a binary blob. While this might be simpler it prevents us from
+being able to push down optimizations into the map.
+
+</td>
+
+<tr>
+<td> MzTimestamp </td>
+<td>
+
+```rust
+struct MzTimestamp(u64);
+```
+
+</td>
+<td>
+
+`PrimitiveArray<u64>`
+
+</td>
+<td>
+
+Number of milliseconds since the UNIX epoch.
+
+</td>
+
+<tr>
+<td> Range </td>
+<td>
+
+```rust
+struct Range<T> {
+  lower: RangeBound<T> {
+    inclusize: bool,
+    bound: Option<T>,
+  },
+  upper: RangeBound<T>,
+}
+```
+
+</td>
+<td>
+
+```
+RangeBound: StructArray<{
+  inclusive: bool,
+  bound: T,
+}>
+```
+
+```
+Range: StructArray<{
+  lower: RangeBound,
+  upper: RangeBound,
+}>
+```
+
+</td>
+<td>
+
+Structure the data as it is in Rust.
+
+Ranges seem pretty interesting and powerful, so Persist having an understanding
+of the data seems worthwhile for the long term. They could also be entirely
+unused (I'm not sure) in which case the complexity of encoding these in a
+structured way might not be worth it.
+
+> **Alternative:** Encode a Range into a binary format and store it as a blob.
+
+</td>
+
+<tr>
+<td> MzAclItem </td>
+<td>
+
+```rust
+struct MzAclItem {
+  // String
+  grantee: RoleId,
+  // String
+  grantor: RoleId,
+  // u64
+  acl_mode: AclMode,
+}
+```
+
+</td>
+<td>
+
+```
+StructArray<{
+  grantee: String,
+  grantor: String,
+  acl_mode: u64,
+}>
+```
+
+</td>
+<td>
+
+Structure the data as it is in Rust.
+
+> **Alternative:** Encode an MzAclItem into a binary format and store it as a blob.
+
+</td>
+
+<tr>
+<td> AclItem </td>
+<td>
+
+```rust
+struct AclItem {
+  // u32
+  grantee: Oid,
+  // u32
+  grantor: Oid,
+  // u64
+  acl_mode: AclMode,
+}
+```
+
+</td>
+<td>
+
+```
+StructArray<{
+  grantee: u32,
+  grantor: u32,
+  acl_mode: u64,
+}>
+```
+
+</td>
+<td>
+
+Structure the data as it is in Rust.
+
+> **Alternative:** It would be relatively easy to stitch together the three
+values that make up an `AclItem` into a `FixedSizeBinary<16>`, it should even
+sort the same as its Rust counterpart.
+
+</td>
+
+<tr>
+<td> Int2Vector </td>
+<td>
+
+```rust
+Vec<i16>
+```
+
+</td>
+<td>
+
+`VariableSizeList<i16>`
+
+</td>
+<td>
+
+Structure the data as it is in Rust.
+
+</td>
+
+</table>
+
+### Extension Types
+
+Arrow has the concept of [Extension Types](https://arrow.apache.org/docs/format/Columnar.html#extension-types)
+which do not change the physical layout of a column, but allow you to identify
+the kind of data stored in a column by tagging it with metadata. At the time
+of writing there are only two accepted [canonical extension types](https://arrow.apache.org/docs/format/CanonicalExtensions.html#official-list)
+but we are free to create our own.
+
+As part of our mapping from `Datum` to Arrow column we could include an
```

Comment:
That's fair! We can shorten this then to something like `mz.p.<v>.<datum type tag>`, e.g. Assuming String has a type tag of 12 it would look like `mz.p.1.12`.

I like having the `Datum` specific version because it allows us to evolve each `Datum` encoding independently which allows us to avoid unnecessary migrations. For example, say we change the encoding of `Interval`, if we bumped the overall format version then all batches regardless of whether they contain an `Interval`, would need to migrate during consolidation. But with a `Datum` specific version we would only have to migrate batches that contain `Interval`s